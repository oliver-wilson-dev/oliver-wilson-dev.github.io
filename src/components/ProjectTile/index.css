:root {
    --marginRight: 20px;
    --boxShadowHOffset: 5px;
}

.projectTile {
    max-width: calc((var(--pageContentMaxWidth)) - 40px - (var(--marginRight) * 2));

    margin: 20px var(--marginRight) 0px var(--marginRight);
    box-shadow: var(--boxShadowHOffset) 5px 12px -2px var(--primaryBorderColour);
    padding: 20px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}

.footer {
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    margin-left: auto;
    margin-right: auto;
    align-items: center;
    justify-content: space-around;
}

.techHeader {
    margin: 20px 0px 0px 0px;
}

.techList {
    font-weight: 600;
}

.buttonLink {
    font-size: 14px;
    text-align: center;
    margin: 7px;
}

@media (min-width: 768px) {
    :root {
        --numberOfTiles: 2;
        --numberOfTilesMinusOne: 1;
    }

    .projectTile {
        max-width: calc((100% / var(--numberOfTiles)) - (var(--marginRight) * var(--numberOfTilesMinusOne)) - (var(--boxShadowHOffset) * var(--numberOfTiles)) - 20px);

        margin: 20px var(--marginRight) 0px 0px;

        &:nth-of-type(2n) {
            margin-right: 0px;
        }
    }
}

@media (min-width: 1024px) {
    :root {
        --numberOfTiles: 3;
        --numberOfTilesMinusOne: 2;
    }

    .projectTile {
        max-width: calc((100% / var(--numberOfTiles)) - (var(--marginRight) * var(--numberOfTilesMinusOne)) - (var(--boxShadowHOffset) * var(--numberOfTiles)));

        &:nth-of-type(2n) {
            margin-right: var(--marginRight);
        }

        &:nth-of-type(3n) {
            margin-right: 0px;
        }
    }
}