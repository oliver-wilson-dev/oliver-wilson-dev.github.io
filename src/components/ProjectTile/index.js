import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './index.css';
import sharedStyles from '../../styles/shared.css';

const ProjectTile = ({
  name,
  codeHref,
  demoHref,
  description,
  technologies
}) => (
  <article className={styles.projectTile}>
    <header>
      <h3>{name}</h3>
      {technologies && (
      <React.Fragment>
        <h4 className={styles.techHeader}>Technologies</h4>
        <p className={styles.techList}>
          {technologies.join(', ')}
        </p>
      </React.Fragment>
      )}
    </header>
    <p>{description}</p>
    {(codeHref || demoHref) && (
      <footer className={styles.footer}>
        {demoHref && <a href={demoHref} className={cn(sharedStyles.buttonLink, styles.buttonLink)} rel="noopener noreferrer" target="_blank">View demo</a>}
        {codeHref && <a href={codeHref} className={cn(sharedStyles.buttonLink, styles.buttonLink)} rel="noopener noreferrer" target="_blank">View code</a>}
      </footer>
    )}
  </article>
);

ProjectTile.propTypes = {
  codeHref: PropTypes.string,
  demoHref: PropTypes.string,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  technologies: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default ProjectTile;
