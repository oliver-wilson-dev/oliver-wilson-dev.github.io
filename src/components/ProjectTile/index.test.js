import React from 'react';
import { shallow } from 'enzyme';
import ProjectTile from '.';

const render = (overrideProps = {}) => shallow(
  <ProjectTile
    name="test-project-name"
    description="test-description"
    technologies={['react', 'redux', 'node']}
    {...overrideProps}
  />
);

describe('ProjectTile', () => {
  describe('when provided with a name, technologies list and description', () => {
    it('should render correctly', () => {
      expect(render()).toMatchSnapshot();
    });

    describe('and a codeHref', () => {
      const overrideProps = {
        codeHref: 'https://test-href',
      };

      it('should render correctly', () => {
        expect(render(overrideProps)).toMatchSnapshot();
      });

      describe('and a demoHref', () => {
        const overrideProps = {
          codeHref: 'https://test-href',
          demoHref: 'https://test-href',
        };

        it('should render correctly', () => {
          expect(render(overrideProps)).toMatchSnapshot();
        });
      });
    });

    describe('and a demoHref', () => {
      const overrideProps = {
        demoHref: 'https://test-href',
      };

      it('should render correctly', () => {
        expect(render(overrideProps)).toMatchSnapshot();
      });
    });
  });
});
