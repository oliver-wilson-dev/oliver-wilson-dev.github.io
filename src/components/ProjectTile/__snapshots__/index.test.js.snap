// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTile when provided with a name, technologies list and description and a codeHref and a demoHref should render correctly 1`] = `
<article
  className="projectTile"
>
  <header>
    <h3>
      test-project-name
    </h3>
    <h4
      className="techHeader"
    >
      Technologies
    </h4>
    <p
      className="techList"
    >
      react, redux, node
    </p>
  </header>
  <p>
    test-description
  </p>
  <footer
    className="footer"
  >
    <a
      className="buttonLink buttonLink"
      href="https://test-href"
      target="_blank"
    >
      View demo
    </a>
    <a
      className="buttonLink buttonLink"
      href="https://test-href"
      target="_blank"
    >
      View code
    </a>
  </footer>
</article>
`;

exports[`ProjectTile when provided with a name, technologies list and description and a codeHref should render correctly 1`] = `
<article
  className="projectTile"
>
  <header>
    <h3>
      test-project-name
    </h3>
    <h4
      className="techHeader"
    >
      Technologies
    </h4>
    <p
      className="techList"
    >
      react, redux, node
    </p>
  </header>
  <p>
    test-description
  </p>
  <footer
    className="footer"
  >
    <a
      className="buttonLink buttonLink"
      href="https://test-href"
      target="_blank"
    >
      View code
    </a>
  </footer>
</article>
`;

exports[`ProjectTile when provided with a name, technologies list and description and a demoHref should render correctly 1`] = `
<article
  className="projectTile"
>
  <header>
    <h3>
      test-project-name
    </h3>
    <h4
      className="techHeader"
    >
      Technologies
    </h4>
    <p
      className="techList"
    >
      react, redux, node
    </p>
  </header>
  <p>
    test-description
  </p>
  <footer
    className="footer"
  >
    <a
      className="buttonLink buttonLink"
      href="https://test-href"
      target="_blank"
    >
      View demo
    </a>
  </footer>
</article>
`;

exports[`ProjectTile when provided with a name, technologies list and description should render correctly 1`] = `
<article
  className="projectTile"
>
  <header>
    <h3>
      test-project-name
    </h3>
    <h4
      className="techHeader"
    >
      Technologies
    </h4>
    <p
      className="techList"
    >
      react, redux, node
    </p>
  </header>
  <p>
    test-description
  </p>
</article>
`;
